### PR TITLE
Redux saga implementation, server side rendering included

### PR DESCRIPTION
**Pull request by IvanJov**
@donovanh @feugy

### CODE REVIEW

File: client/src/store/sagas/app.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { delay } from 'redux-saga'
+import { put, takeEvery } from 'redux-saga/effects'
+import {SET_MESSAGE, SET_MESSAGE_FINAL} from '../constants/app'
+
+export function* appSaga(message) {
+  // wait for 2 seconds and dispatch the new action
+  yield delay(2000)
```
Comment (donovanh) [inline client/src/store/sagas/app.js]: Cool to see this in action. Maybe not needed though for production?

Comment (IvanJov) [reply client/src/store/sagas/app.js]: This is just an example of the sagas, something like an async call that we need to wait for. We'll remove this one when we make our first page with some real API call 😄

**Comment by Joezo** [comment]
I've just stumbled across this whilst looking through everything and I have a couple of questions.

Why sagas over thunks? Do we have a specific need for them?

Do we have to use generator functions over async/await?

I'm asking all these things because in the past i've only really used thunks & async/await with redux and found it a good balance of simple/powerful. Here's an example of the way i've been doing it. 

https://gist.github.com/Joezo/77ad66836a8b6892b395cabc3508df54

I see you've just added standardjs, can we *please* use prettier? it's so much easier.

**Comment by donovanh** [comment]
I have no preference on standard vs prettier - though I find standard pretty easy to use as it's easy to enforce via both my editor and when adding commits. Not dogmatic though.

On the sagas / thunks - I've only ever used vanilla Redux actions and reducers so am not sure of the benefits of sagas. I'd be interested in your thoughts @IvanJov. I guess it might be worth sticking with as simple an approach as we can until they're needed.

**Comment by Joezo** [comment]
My preference for prettier is based solely on it's ability to fix **everything** on save whereas I know standard gets stuck with some things. I don't care what rules we have on/off as long as hitting cmd + s fixes them all and I don't have to ever concern myself with them 👍 

On your piece about sagas/thunks, that was my thought, it may be easy to keep it simple, although i'm keen to hear from @IvanJov on it.

**Comment by IvanJov** [comment]
@Joezo Hey 👋
Regarding all your questions, we had a discussion what are we going to use, so Sagas/StandardJS won 😄But you know, we can always change it.

Regarding Sagas, I am a big fan of it because it's just appending some functionality on top of the Redux. In the previous company, we had issues with thunk and that actions promise approach. In my opinion, and regarding redux documentation, actions should be pure functions that return objects. And then on top of that, you add sagas that listen for actions and do some async side effects and again dispatch pure objects when side effects are done.

The same thing goes with redux-observables and RxJS (that was another option we discussed), although that requires some knowledge about reactive programming and I didn't want to force people to spend time learning it 😄

We can always go back to thunk, but I'd personally avoid it.

**Comment by Joezo** [comment]
hey @IvanJov 

Yup I think i'm around 80% of the way there to knowing how sagas work. What particular pieces of this project do you think they'll help towards over thunks? Honestly, it looks like a lot of boilerplate (even more than standard redux 😂) that splits out related code for the sake of "pure actions".

I'm just trying to see how using them makes writing this project easier. I'm not saying we should use my little example up there, but I can quite easily go into a "module" file like it and see everything to do with that area of the state. How it gets there, what requests are sent, how we deal with results etc. I'm not hunting around for pieces of code that might run when I dispatch an action and ultimately losing time. 

I'm happy to go with Sagas, I just want to make sure we all understand how and **why** we're using them.

**Comment by IvanJov** [comment]
@Joezo 
Hmm not sure for some particular pieces, but basically everything async will go through sagas. For example, we'll probably save our logged in user in redux store, or we'll save JWT token. When you want for example to fetch users favorite products from the backend, you'll just need to dispatch "FETCH_FAVORITE_PRODUCTS" action, for example, saga will read from redux, fetch from API, do whatever operation you need and store the data in the redux by dispatching some save action. React will then re-render the page and favorite products will be shown. Saga is basically removing all logic from actions and living as a thing on the side that does the magic when you dispatch some actions.

And in my opinion, that's more readable because all the business logic is living in sagas. Everything from reading from the store, updating it, sending async requests, showing some loading, progress, whatever, all those things are on the side.

One more thing, API canceling works well with sagas. If you, for example, select some filter on the product page, we'll send API request to get those products. If you select one more, you'll have one more API request, one for one filter, second for both, and so on. With sagas you can cancel previous ones and make the app more performant because it will wait for one request at the time. And so on, really sagas give you so many things 😎

And dealing with the results is also simple, it always goes back to the store from saga and component can read it.

**Comment by IvanJov** [comment]
@Joezo This can be helpful, it lists all saga effects and helpers https://redux-saga.js.org/docs/api/ 😄

**Comment by Joezo** [comment]
Sounds like it does everything that thunks does which is cool. The cancelling bit is nice. Know of any guides to get up to speed on what saga effects to use in what situations etc? A cheat sheet perhaps.

Happy for this to be merged now, just wanted to understand the thought behind it. Assuming @donovanh and @feugy are happy with the code, I haven't looked at it extensively or even tried to run it yet.

**Comment by donovanh** [comment]
LGTM - let me know if I can help with any conflicts on merge

**Pull request merged**